### ISSUE TITLE
SBT ignores alternative scala distributions when deciding to use the same classloader as the build

### ISSUE THREAD

**Issue by xeno-by** [sbt/sbt#1172]
Minimized reproduction: https://github.com/xeno-by/testvirt

**Comment by jsuereth** [comment]
Thanks for the report.   So, if I were to rename this it would be:

`Sbt refuses to use releases of non-(org.scala-lang) Scala distros if it matches the version of an (org.scala-lang) distro`?

We'll have to check the update report to see if somehow 2.10.2 is leaking onto your classpath at an earlier spot than scala-virtualized.

**Comment by xeno-by** [comment]
Well, actually there is 2.10.2-RC2 for vanilla Scala, so it's not like the problem is in a version clash.

**Comment by namin** [comment]
I can't reproduce the issue! `sbt run` works fine for me. `sbt --version` returns 0.13.1.

**Comment by jsuereth** [comment]
@xeno-by is on windows still?.  We already know there's some classpath ordering issues there (even with Mac).

My guess is that somehow both scala's are resolved in your project and the wrong one ends up in the right place at the right time.   But I won't know until I have a chance to reproduce locally (I'm still running regression tests now, so my machine is fully loaded).

**Comment by xeno-by** [comment]
I'm on Mac. Using Paul's sbt runner.

**Comment by jsuereth** [comment]
Can you throw a `project/build.properties` in there?  That's becoming a requirement for sbt.

**Comment by cunei** [comment]
Here I'm getting:

```
cunei@rover:~/activities/triaging/sbt/1172/testvirt$ sbt compile
Loading /home/cunei/activities/sbt-initialtests/13-1/sbt/bin/sbt-launch-lib.bash
[warn] The global sbt directory is now versioned and is located at /home/cunei/.sbt/0.13.
[warn]   You are seeing this warning because there is global configuration in /home/cunei/.sbt but not in /home/cunei/.sbt/0.13.
[warn]   The global sbt directory may be changed via the sbt.global.base system property.
[info] Set current project to paradisevirt (in build file:/home/cunei/activities/triaging/sbt/1172/testvirt/)
[info] Updating {file:/home/cunei/activities/triaging/sbt/1172/testvirt/}testvirt...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] downloading http://repo1.maven.org/maven2/org/scala-lang/virtualized/scala-compiler/2.10.2/scala-compiler-2.10.2.jar ...
[info]  [SUCCESSFUL ] org.scala-lang.virtualized#scala-compiler;2.10.2!scala-compiler.jar (28245ms)
[info] downloading http://repo1.maven.org/maven2/org/scala-lang/virtualized/scala-library/2.10.2/scala-library-2.10.2.jar ...
[info]  [SUCCESSFUL ] org.scala-lang.virtualized#scala-library;2.10.2!scala-library.jar (9808ms)
[info] downloading http://repo1.maven.org/maven2/org/scala-lang/virtualized/scala-reflect/2.10.2/scala-reflect-2.10.2.jar ...
[info]  [SUCCESSFUL ] org.scala-lang.virtualized#scala-reflect;2.10.2!scala-reflect.jar (5039ms)
[info] downloading http://repo1.maven.org/maven2/org/scala-lang/virtualized/jline/2.10.2/jline-2.10.2.jar ...
[info]  [SUCCESSFUL ] org.scala-lang.virtualized#jline;2.10.2!jline.jar (1012ms)
[info] Done updating.
[info] Compiling 1 Scala source to /home/cunei/activities/triaging/sbt/1172/testvirt/target/scala-2.10/classes...
[info] 'compiler-interface' not yet compiled for Scala 2.10.2-20140310-140650-2481f036a5. Compiling...
[info]   Compilation completed in 11.679 s
[warn] there were 1 feature warning(s); re-run with -feature for details
[warn] one warning found
[success] Total time: 62 s, completed Mar 11, 2014 2:30:38 PM
```

**Comment by xeno-by** [comment]
Done. Forced version to 0.13.1, and that didn't help.

**Comment by xeno-by** [comment]
Weird, for me it says `'compiler-interface' not yet compiled for Scala
2.10.2` :(

**Comment by xeno-by** [comment]
Cleaning .ivy2 cache doesn't help either.

**Comment by jsuereth** [comment]
@xeno-by What if you use sbt-launcher directly.   It still looks to me like both scala's are on your scala-instance classpath.  Can you dump your `sbt -d` output?

**Comment by namin** [comment]
@xeno-by -- Does 2.10.1 (instead of 2.10.2) work for you? In the past, I noticed that it takes a few hours for a new version to be available everywhere (Scala-Virtualized 2.10.2 was just released this morning...).

**Comment by xeno-by** [comment]
It downloads 2.10.2 virt successfully, and I checked the ivy cache to make
sure it contains the right jar.

**Comment by xeno-by** [comment]
Here's `sbt -d` from the rebel cut:

14:40 ~/Projects/testvirt (master)$ sbt -d
No extra sbt options have been defined
Detected sbt version 0.13.1
Starting sbt-the-rebel-cut: invoke with -help for other options
Using /Users/xeno_by/.sbt/0.13.1 as sbt dir, -sbt-dir to override.
[addJava] arg = '-Dsbt.global.base=/Users/xeno_by/.sbt/0.13.1'
Using default jvm options

# Executing command line:

/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home/bin/java
-Dfile.encoding=UTF8
-XX:MaxPermSize=384m
-Xms512m
-Xmx1536m
-Xss2m
-XX:ReservedCodeCacheSize=256m
-XX:+TieredCompilation
-XX:+CMSClassUnloadingEnabled
-XX:+UseConcMarkSweepGC
-Dsbt.global.base=/Users/xeno_by/.sbt/0.13.1
-jar
/Users/xeno_by/.sbt/launchers/0.13.1/sbt-launch.jar
shell

[info] Loading project definition from
/Users/xeno_by/Projects/testvirt/project
[info] Updating
{file:/Users/xeno_by/Projects/testvirt/project/}testvirt-build...
Waiting for lock on /Users/xeno_by/.ivy2/.sbt.ivy.lock to be available...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Set current project to paradisevirt (in build
file:/Users/xeno_by/Projects/testvirt/)

> compile
> [info] Compiling 1 Scala source to
> /Users/xeno_by/Projects/testvirt/target/scala-2.10/classes...
> Waiting for lock on
> /Users/xeno_by/.sbt/0.13.1/boot/scala-2.10.3/org.scala-sbt/sbt/0.13.1/sbt.components.lock
> to be available...
> [error] /Users/xeno_by/Projects/testvirt/Test.scala:5: value makeNewVar is
> not a member of scala.tools.nsc.ast.parser.TreeBuilder
> [error]   def test = builder.makeNewVar(???)
> [error]                      ^
> [error] one error found
> [error](compile:compile) Compilation failed
> [error] Total time: 19 s, completed Mar 11, 2014 2:40:53 PM

Let me try the launcher directly.

**Comment by xeno-by** [comment]
Downloaded the launcher from http://www.scala-sbt.org/release/docs/Getting-Started/Setup.html. Didn't help. Here's the output:

```
14:42 ~/Projects/testvirt (master)$ ./sbt/bin/sbt -d
Loading /Users/xeno_by/Projects/testvirt/sbt/bin/sbt-launch-lib.bash
# Executing command line:
java
-Xms1024m
-Xmx1024m
-XX:MaxPermSize=256m
-XX:ReservedCodeCacheSize=128m
-jar
/Users/xeno_by/Projects/testvirt/sbt/bin/sbt-launch.jar

[info] Loading project definition from /Users/xeno_by/Projects/testvirt/project
[info] Set current project to paradisevirt (in build file:/Users/xeno_by/Projects/testvirt/)
> compile
[info] Compiling 1 Scala source to /Users/xeno_by/Projects/testvirt/target/scala-2.10/classes...
[info] 'compiler-interface' not yet compiled for Scala 2.10.2. Compiling...
[info]   Compilation completed in 10.708 s
[error] /Users/xeno_by/Projects/testvirt/Test.scala:5: value makeNewVar is not a member of scala.tools.nsc.ast.parser.TreeBuilder
[error]   def test = builder.makeNewVar(???)
[error]                      ^
[error] one error found
[error] (compile:compile) Compilation failed
[error] Total time: 12 s, completed Mar 11, 2014 2:43:11 PM
>
```

**Comment by jsuereth** [comment]
Hmm...

1) Can you show me what's in your `~/.sbt/boot/scala-2.10.3/org.scala-sbt/sbt/0.13.1/` directory?
  Specifically the compiler-interface-bin_\* directories
2) Try `reboot full` (clears the launcher/incremental compiler interface cache) in the sbt console.

**Comment by cunei** [comment]
Also, what does it show if you try `show fullClasspath` ?

**Comment by xeno-by** [comment]
```
14:52 ~/Projects/testvirt (master)$ sbt/bin/sbt
Loading /Users/xeno_by/Projects/testvirt/sbt/bin/sbt-launch-lib.bash
[info] Loading project definition from /Users/xeno_by/Projects/testvirt/project
[info] Set current project to paradisevirt (in build file:/Users/xeno_by/Projects/testvirt/)
> reboot full
Getting org.scala-sbt sbt 0.13.1 ...
:: retrieving :: org.scala-sbt#boot-app
    confs: [default]
    43 artifacts copied, 0 already retrieved (12646kB/51ms)
Getting Scala 2.10.3 (for sbt)...
:: retrieving :: org.scala-sbt#boot-scala
    confs: [default]
    5 artifacts copied, 0 already retrieved (24447kB/47ms)
[info] Loading project definition from /Users/xeno_by/Projects/testvirt/project
[info] Set current project to paradisevirt (in build file:/Users/xeno_by/Projects/testvirt/)
> compile
[info] Compiling 1 Scala source to /Users/xeno_by/Projects/testvirt/target/scala-2.10/classes...
[error] /Users/xeno_by/Projects/testvirt/Test.scala:5: value makeNewVar is not a member of scala.tools.nsc.ast.parser.TreeBuilder
[error]   def test = builder.makeNewVar(???)
[error]                      ^
[error] one error found
[error] (compile:compile) Compilation failed
[error] Total time: 2 s, completed Mar 11, 2014 2:52:56 PM
> 14:52 ~/Projects/testvirt (master)$  ~/.sbt/boot/scala-2.10.3/org.scala-sbt/sbt/0.13.1/
bash: /Users/xeno_by/.sbt/boot/scala-2.10.3/org.scala-sbt/sbt/0.13.1/: Is a directory
14:53 ~/Projects/testvirt (master)$ ls ~/.sbt/boot/scala-2.10.3/org.scala-sbt/sbt/0.13.1/
actions-0.13.1.jar          ivy-2.3.0-rc1.jar
api-0.13.1.jar              jline-2.11.jar
apply-macro-0.13.1.jar          jsch-0.1.46.jar
cache-0.13.1.jar            launcher-interface-0.13.1.jar
classfile-0.13.1.jar            logging-0.13.1.jar
classpath-0.13.1.jar            main-0.13.1.jar
collections-0.13.1.jar          main-settings-0.13.1.jar
command-0.13.1.jar          persist-0.13.1.jar
compile-0.13.1.jar          process-0.13.1.jar
compiler-integration-0.13.1.jar     relation-0.13.1.jar
compiler-interface-bin_2.10.2__50.0 run-0.13.1.jar
compiler-interface-bin_2.10.3       sbinary_2.10-0.4.2.jar
compiler-interface-bin_2.8.2.final  sbt-0.13.1.jar
compiler-interface-bin_2.9.2        sbt.components.lock
compiler-interface-bin_2.9.3        scala-reflect-2.10.3.jar
compiler-interface-src          task-system-0.13.1.jar
compiler-ivy-integration-0.13.1.jar tasks-0.13.1.jar
completion-0.13.1.jar           test-agent-0.13.1.jar
control-0.13.1.jar          test-interface-1.0.jar
cross-0.13.1.jar            testing-0.13.1.jar
incremental-compiler-0.13.1.jar     tracking-0.13.1.jar
io-0.13.1.jar               xsbti
ivy-0.13.1.jar
14:53 ~/Projects/testvirt (master)$
```

**Comment by xeno-by** [comment]
```
> show fullClasspath
[info] Compiling 1 Scala source to /Users/xeno_by/Projects/testvirt/target/scala-2.10/classes...
[error] /Users/xeno_by/Projects/testvirt/Test.scala:5: value makeNewVar is not a member of scala.tools.nsc.ast.parser.TreeBuilder
[error]   def test = builder.makeNewVar(???)
[error]                      ^
[error] one error found
[error] (compile:compile) Compilation failed
[error] Total time: 2 s, completed Mar 11, 2014 2:53:42 PM
```

**Comment by cunei** [comment]
Fair enough. Let's say `show dependencyClasspath`, instead.

**Comment by cunei** [comment]
Could it be any global plugin?

**Comment by xeno-by** [comment]
```
> show fullClasspath
[info] Compiling 1 Scala source to /Users/xeno_by/Projects/testvirt/target/scala-2.10/classes...
[error] /Users/xeno_by/Projects/testvirt/Test.scala:5: value makeNewVar is not a member of scala.tools.nsc.ast.parser.TreeBuilder
[error]   def test = builder.makeNewVar(???)
[error]                      ^
[error] one error found
[error] (compile:compile) Compilation failed
[error] Total time: 2 s, completed Mar 11, 2014 3:53:41 PM
> show dependencyClasspath
[info] List(Attributed(/Users/xeno_by/.sbt/0.13.0/boot/scala-2.10.2/lib/scala-compiler.jar), Attributed(/Users/xeno_by/.sbt/0.13.0/boot/scala-2.10.2/lib/scala-library.jar), Attributed(/Users/xeno_by/.sbt/0.13.0/boot/scala-2.10.2/lib/scala-reflect.jar))
[success] Total time: 0 s, completed Mar 11, 2014 3:53:58 PM
```

**Comment by xeno-by** [comment]
How do I see whether I have any global plugins?

**Comment by jsuereth** [comment]
Depends on your script.  For normal sbt, check `~/.sbt/0.13`. For paul's script, check `~/.sbt/<version>`

**Comment by xeno-by** [comment]
```
15:58 ~/Projects/testvirt (master)$ ls ~/.sbt/0.13
staging
15:58 ~/Projects/testvirt (master)$ ls ~/.sbt/0.13.1
boot    staging target
```

**Comment by cunei** [comment]
My classpath is:

```
> show dependencyClasspath
[info] List(Attributed(/home/cunei/.ivy2/cache/org.scala-lang.virtualized/scala-compiler/jars/scala-compiler-2.10.2.jar), Attributed(/home/cunei/.ivy2/cache/org.scala-lang.virtualized/scala-library/jars/scala-library-2.10.2.jar), Attributed(/home/cunei/.ivy2/cache/org.scala-lang.virtualized/scala-reflect/jars/scala-reflect-2.10.2.jar))
```

Which seems somewhat more sensible. Somehow yours is picking Scala from the sbt boot dir. Not sure why. Do you have anything sbt-related in your env?

**Comment by cunei** [comment]
Can you please try moving out of the way `/Users/xeno_by/.sbt/0.13.0` ?

**Comment by xeno-by** [comment]
I removed the `.sbt` directory altogether, and now trying to launch sbt produces

```
> compile
[trace] Stack trace suppressed: run last compile:compile for the full output.
[error] (compile:compile) sbt.InvalidScalaInstance: Scala instance doesn't exist or is invalid:
[error]     version 2.10.2, library jar: /Users/xeno_by/.sbt/0.13.0/boot/scala-2.10.2/lib/scala-library.jar, compiler jar: /Users/xeno_by/.sbt/0.13.0/boot/scala-2.10.2/lib/scala-compiler.jar
[error] Total time: 0 s, completed Mar 11, 2014 4:24:10 PM
```

**Comment by xeno-by** [comment]
What do you mean by sbt-related in my env? What should I check for?

**Comment by cunei** [comment]
SBT_OPTS, for instance

**Comment by xeno-by** [comment]
```
16:29 ~$ echo $SBT_OPTS

```

**Comment by cunei** [comment]
There has to be _some_ reason why it is picking the boot dir of a different sbt version. :) Can you please add `set -x` at the beginning of `sbt/bin/sbt`, and relaunch? I'm curious to see the final invocation of java. Also, can you please try in a completely empty new dir:

```
git clone git@github.com:xeno-by/testvirt.git
cd testvirt
mkdir 0131 ; wget -O - http://repo.scala-sbt.org/scalasbt/sbt-native-packages/org/scala-sbt/sbt/0.13.1/sbt.tgz | tar xfz - -C 0131
./0131/sbt/bin/sbt
```

Since you also removed `.sbt`, there should be really nothing around that can interfere any longer.

**Comment by xeno-by** [comment]
Amazing... It now works!

```
16:43 ~$ prj
16:43 ~/Projects$ rm -rf testvirt/
16:43 ~/Projects$ git clone git@github.com:xeno-by/testvirt.git
Cloning into 'testvirt'...
remote: Counting objects: 14, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 14 (delta 4), reused 11 (delta 2)
Receiving objects: 100% (14/14), done.
Resolving deltas: 100% (4/4), done.
Checking connectivity... done.
16:43 ~/Projects$ cd testvirt
16:43 ~/Projects/testvirt (master)$ mkdir 0131 ; wget -O - http://repo.scala-sbt.org/scalasbt/sbt-native-packages/org/scala-sbt/sbt/0.13.1/sbt.tgz | tar xfz - -C 0131
--2014-03-11 16:43:30--  http://repo.scala-sbt.org/scalasbt/sbt-native-packages/org/scala-sbt/sbt/0.13.1/sbt.tgz
Resolving repo.scala-sbt.org (repo.scala-sbt.org)... 107.23.26.55, 107.23.74.195, 107.23.88.175
Connecting to repo.scala-sbt.org (repo.scala-sbt.org)|107.23.26.55|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 994631 (971K) [application/x-gzip]
Saving to: ‘STDOUT’

100%[==========================================================>] 994,631      318KB/s   in 3.1s

2014-03-11 16:43:34 (318 KB/s) - written to stdout [994631/994631]

16:43 ~/Projects/testvirt (master)$ ./0131/sbt/bin/sbt
Loading /Users/xeno_by/Projects/testvirt/0131/sbt/bin/sbt-launch-lib.bash
[info] Loading project definition from /Users/xeno_by/Projects/testvirt/project
[info] Updating {file:/Users/xeno_by/Projects/testvirt/project/}testvirt-build...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Set current project to paradisevirt (in build file:/Users/xeno_by/Projects/testvirt/)
> compile
[info] Updating {file:/Users/xeno_by/Projects/testvirt/}testvirt...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] downloading http://repo1.maven.org/maven2/org/scala-lang/virtualized/scala-compiler/2.10.2/scala-compiler-2.10.2.jar ...
[info]  [SUCCESSFUL ] org.scala-lang.virtualized#scala-compiler;2.10.2!scala-compiler.jar (2993ms)
[info] downloading http://repo1.maven.org/maven2/org/scala-lang/virtualized/scala-library/2.10.2/scala-library-2.10.2.jar ...
[info]  [SUCCESSFUL ] org.scala-lang.virtualized#scala-library;2.10.2!scala-library.jar (1623ms)
[info] downloading http://repo1.maven.org/maven2/org/scala-lang/virtualized/scala-reflect/2.10.2/scala-reflect-2.10.2.jar ...
[info]  [SUCCESSFUL ] org.scala-lang.virtualized#scala-reflect;2.10.2!scala-reflect.jar (975ms)
[info] downloading http://repo1.maven.org/maven2/org/scala-lang/virtualized/jline/2.10.2/jline-2.10.2.jar ...
[info]  [SUCCESSFUL ] org.scala-lang.virtualized#jline;2.10.2!jline.jar (337ms)
[info] Done updating.
[info] Compiling 1 Scala source to /Users/xeno_by/Projects/testvirt/target/scala-2.10/classes...
[info] 'compiler-interface' not yet compiled for Scala 2.10.2-20140310-140650-2481f036a5. Compiling...
[info]   Compilation completed in 11.101 s
[warn] there were 1 feature warning(s); re-run with -feature for details
[warn] one warning found
[success] Total time: 22 s, completed Mar 11, 2014 4:44:15 PM
```

**Comment by cunei** [comment]
There must have been some junk left over in the project's local `target` or `project` dirs. Did you maybe save it beforehand?

**Comment by xeno-by** [comment]
Unfortunately I just did `rm -rf` :(

**Comment by xeno-by** [comment]
But the same problem is reproducible on Jenkins: https://scala-webapps.epfl.ch/jenkins/view/misc/job/macro-paradise-nightly-2.10.2-virt/14/console

**Comment by xeno-by** [comment]
I'm rerunning the build there, so that you'll soon be able to take a look at the artifacts: https://scala-webapps.epfl.ch/jenkins/view/misc/job/macro-paradise-nightly-2.10.2-virt/16/console.

**Comment by cunei** [comment]
I cannot get to that machine, but if you can tgz the build dir we can (maybe) have a look later. Since a cleanup solves the problem, the ticket for the moment can be closed. Cheers! :)

**Comment by xeno-by** [comment]
Thank you!

**Comment by jsuereth** [comment]
I think we may still have a legitimate issue, see: https://github.com/sbt/sbt/blob/0.13/main/src/main/scala/sbt/Defaults.scala#L321

**Comment by jsuereth** [comment]
Hey @xeno-by I just got around to fixing this bug, but I can't accurately detect if I've fixed it.  Are you and @namin still working scala-virtualized?  Will there be a 2.10.4 release?  If not, I may have wait on pushing a fix.

Thanks!
- Josh

**Comment by namin** [comment]
Hi Josh,

There's no immediate plan for a release, but if you'd like, I could do one?

Within the next weeks, we'll be looking into whether it's possible to support the most important features of Scala-Virtualized without a custom compiler, relying on macros instead.

**Comment by jsuereth** [comment]
It's not important _unless_ you guys are still dealing with this bug.  We had committed to having this bug fixed for 0.13.6 which is shipping soon, but I have no way to tell if I fixed it, because now that the scala versions are mismatched, we don't run into it ;).

So yeah, if you don't think you need this bug fixed in the future, please close the issue.  Otherwise, let me know if you have another reproduction, as I have a (hopeful) fix sitting in a local branch.

**Comment by namin** [comment]
I tried a few combinations of things but cannot reproduce the bug right now, even though I've seen it happen once or twice again when upgrading some projects to use the latest Scala-Virtualized. Given that adding the build.properties file and removing all target directories seems enough to fix the problem when it occurs, I don't mind the status quo. So, feel free to close this issue (I don't have the permission to do it myself).
